### PR TITLE
Make the rust hyper client `Send` so it can be used in rust threads more easily

### PR DESCRIPTION
**Pull request by blazzy**
The hyper client cannot be used easily in threads as it isn't `Send` friendly. Something like this will fail to compile:

```rust
use petstore_hyper::apis::client::APIClient;
use petstore_hyper::apis::configuration::Configuration;

#[tokio::main]
async fn main() {
    tokio::spawn(async move {
        let client = APIClient::new(Configuration::new());
        let _result = client.pet_api().get_pet_by_id(5).await;
    });
}
```

Issue with more detailed steps to reproduce: https://github.com/OpenAPITools/openapi-generator/issues/19208

This PR makes the hyper client and its internals `Send`. So the client can be used in threads.

It ended up being a little larger of a PR in the pursuit of writing a test. It turns out the tests aren't actually being run when invoking verify on the petstore samples. This PR had to touch a few other thing to ensure `cargo test` would work. Let me know if I should break it up into multiple PRs.

Should I update the pom.xml file to do a `cargo test` instead of a `cargo check`? I'm not sure if that was intentional.

#### Commit summary

This might be a good order to review the changes:

7ee4bcafa044b48dd0fe251b490716a51486ee84 - This is the commit that actually makes the code work in threads. I adds the `Send` constraint to a bunch generic parameters and traits and uses an `Arc` instead of an `Rc`

bda9b6d5f4d4adff0cc7315c329117dfe53b9993 - This commit adds a test to validate that it works and doesn't regress.

e53c04c133a5eb5bf695e3c56fce070e437e3482 - This commit regenerates the samples

The remaining commits were trying to get the tests to run in `samples/client/others/rust/` to run successfully. There were several issues to fix to get tests to run successfully.

43ba4f022943646922e2cd1ded8540184132ea2b 46d3520a5f6447b7d7d9e800e7965e81d17b43fa These set it up so the doc strings are compilable. The first one adds a property called `externCrateName` cribbing from the rust-server generator. The second one uses the externCrateName and fixes other issues with the example rustdoc examples.

ddb330c5e487c39db9972f9ec7d7d5e73fbd4071 The reqwest tests were not compiling and needed a small change.

efbf44503ebb1d76eadea2ff36ca9b871ce3ae47 This changes the petstore swagger file and generates the petstore tests to use https. http://petstore.swagger.io/v2 is 301 redirecting to https://petstore.swagger.io/v2 and this is breaking POSTS.
When the client recieves a redirect it does not resend the POST data, instead it switches to GET. This is in line with how browsers behave when encountering a 301 redirect on a POST request, so I wouldn't consider it a bug in the client, but instead a bug in the test code.

Pinging the rust technical committee members: @frol @farcaller @richardwhiuk @paladinzh @jacob-pro

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java [hunk 1]
```diff
@@ -66,12 +67,13 @@ public class RustClientCodegen extends AbstractRustCodegen implements CodegenCon
...
     public static final String AVOID_BOXED_MODELS = "avoidBoxedModels";
 
-    @Setter protected String packageName = "openapi";
+    protected String packageName = "openapi";
     @Setter protected String packageVersion = "1.0.0";
     protected String apiDocPath = "docs/";
     protected String modelDocPath = "docs/";
     protected String apiFolder = "src/apis";
     protected String modelFolder = "src/models";
+    private String externCrateName = "openapi";
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java:76]: instead of adding another variable, what about simply create an function called `getCrateName()` and we can keep the @Setter above?

Comment (blazzy) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustClientCodegen.java]: Makes sense. Why add more state. I was just blindly copying from the rust server code gen.

Updated in the latest commit: https://github.com/OpenAPITools/openapi-generator/pull/19207/commits/5f8bda343de8361aaf763484961c5716833d1ce1

**Comment by blazzy** [comment]
Added one more commit that makes the api structs `Sync` too also helpful for accessing the data with threads.

**Comment by wing328** [comment]
thanks for the PR. Is it correct to say that this PR is a breaking change? (we do allow breaking changes but need to roll out carefully)

**Comment by blazzy** [comment]
I'd imagine if the user is just importing `APIClient` and `Configuration` and just using those they would be fine.

But if they were importing the different traits and implementing the `trait`s for some reason they would run into compatibility issues. Also if they were importing the individual API structs as opposed instantiating them through the `APIClient` functions they would need to pass an Arc instead of an Rc.

So it's technically not backwards compatible. Though _maybe_ backwards compatible for the most common expected use cases?

I'm not actually sure why the traits exist to be honest. It seems to add a little more complexity for a pay off that is mysterious to me at the moment.

**Comment by blazzy** [comment]
> I'm not actually sure why the traits exist to be honest. It seems to add a little more complexity for a pay off that is mysterious to me at the moment.

Oh. I guess it's to mask the generic parameters for the individual api structs by wrapping them in `Box<dyn trait>`. That makes the APIClient a lot more pleasant to use.

**Comment by wing328** [comment]
can you please resolve the merge conflicts when you've time? i'll try to get it merged for the upcoming release

File: modules/openapi-generator/src/test/resources/3_0/rust/petstore.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 openapi: 3.0.0
 servers:
-  - url: 'http://petstore.swagger.io/v2'
+  - url: 'https://petstore.swagger.io/v2'
```
Comment (wing328) [inline modules/openapi-generator/src/test/resources/3_0/rust/petstore.yaml:3]: please revert the change.

the petstore server should be run locally per https://github.com/OpenAPITools/openapi-generator/wiki/Integration-Tests#how-to-add-integration-tests-for-new-petstore-samples

**Comment by wing328** [comment]
merged via https://github.com/OpenAPITools/openapi-generator/pull/19375

(tested locally with local petstore server)

thanks again for the PR

**Comment by blazzy** [comment]
Thanks! Sorry I didn't have time to swing back to it!
